### ISSUE TITLE
JavadocMethod to allow empty @return tags

### ISSUE THREAD

**Issue by tosix1988** [checkstyle/checkstyle#10014]
Hello,

at the moment, I am not able to configure the JavadocMethod check to allow empty @return tags. My setup:

```
/var/tmp $ cat config.xml
<module name = "Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="allowMissingParamTags" value="false"/>
            <property name="allowMissingReturnTag" value="false"/>
        </module>
    </module>
</module>

/var/tmp $ javac Test.java

/var/tmp $ cat Test.java
public class Test {

	/**
	 * Returns the sum of the two numbers, {@code a} and {@code b}.
	 * 
	 * @param a first summand
	 * @param b
	 * @return
	 */
	public int sum(int a, int b) {
		return a + b;
	}
	
	/**
	 * Returns the product of two numbers, {@code a} and {@code b}.
	 * 
	 * @param a first multiplicand
	 * @param b
	 * @return product of {@code a} and {@code b}
	 */
	public int prod(int a, int b) {
		return a * b;
	}

}
```
Running the checkstyle produces:
```
[WARN] 12: @return tag should be present and have description. [JavadocMethod]
```
While it seems that empty @param tags do not cause any issues, empty @return tags do. Is there a way to configure checkstyle so that empty @return tags are considered valid? I tried using `SuppressionSingleFilter` and filter by message, but the problem with this approach is that missing return tag and empty return tag produce the same message, and I only want to report missing return tags.

If this is not possible, would you consider adding this feature?

**Comment by nmancus1** [comment]
Hello @tosix1988 , thanks for your report. After looking into this a bit, I have a few thoughts:

> While it seems that empty @param tags do not cause any issues...

This is not entirely accurate; in your example, your `@param` tag is not empty:


```bash

➜  src cat Test.java                                 
public class Test {

    /**
     * Returns the sum of the two numbers, {@code a} and {@code b}.
     *
     * @param a first summand
     * @param
     * @return
     */
    public int sum(int a, int b) {
        return a + b;
    }

    /**
     * Returns the product of two numbers, {@code a} and {@code b}.
     *
     * @param a first multiplicand
     * @param b
     * @return product of {@code a} and {@code b}
     */
    public int prod(int a, int b) {
        return a * b;
    }
}

➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="allowMissingParamTags" value="false"/>
            <property name="allowMissingReturnTag" value="false"/>
        </module>
    </module>
</module>

➜  src java -jar checkstyle-8.42-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/Desktop/Tester/src/Test.java:10: @return tag should be present and have description. [JavadocMethod]
[ERROR] /home/nick/Desktop/Tester/src/Test.java:10:31: Expected @param tag for 'b'. [JavadocMethod]
Audit done.
Checkstyle ends with 2 errors.


```

The issue is with how Checkstyle parses Javadoc comments; it appears that Checkstyle Javadoc grammar does not allow for empty tags: 

```bash

➜  src cat Test.java
public class Test {

    /**
     * Returns the sum of the two numbers, {@code a} and {@code b}.
     *
     * @param a first summand
     * @param
     * @return
     */
    public int sum(int a, int b) {
        return a + b;
    }

    /**
     * Returns the product of two numbers, {@code a} and {@code b}.
     *
     * @param a first multiplicand
     * @param b
     * @return product of {@code a} and {@code b}
     */
    public int prod(int a, int b) {
        return a * b;
    }
}

➜  src java -jar checkstyle-8.42-all.jar -J Test.java
Exception in thread "main" java.lang.IllegalArgumentException: [ERROR:9] Javadoc comment at column 0 has parse error. Details: no viable alternative at input '     *' while parsing JAVADOC_TAG
        at com.puppycrawl.tools.checkstyle.DetailNodeTreeStringPrinter.parseJavadocAsDetailNode(DetailNodeTreeStringPrinter.java:71)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.parseAndPrintJavadocTree(AstTreeStringPrinter.java:116)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:97)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:101)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:101)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:101)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:101)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:101)
        at com.puppycrawl.tools.checkstyle.AstTreeStringPrinter.printJavaAndJavadocTree(AstTreeStringPrinter.java:79)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:306)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:191)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:126)

```

This issue was also noticed at https://github.com/checkstyle/checkstyle/pull/9699#discussion_r628346759

It would seem that as far as JavadocMethod is concerned, these tags do not exist because of the parsing error.

**Comment by nmancus1** [comment]
After looking into this, the problem lies in our regex to match javadoc tags:

```java

    /** Compiled regexp to match Javadoc tags with no argument. */
    private static final Pattern MATCH_JAVADOC_NOARG =
            CommonUtil.createPattern("^\\s*(?>\\*|\\/\\*\\*)?\\s*@(return|see)\\s+\\S");

```

This pattern is not matching the empty `@return` tags, and therefore they are not being added to our list of tags.

**Comment by tosix1988** [comment]
Thanks @nmancus1 for your analysis and time.

I thought actually that it is a feature to report empty return tags, judging from the message, which says: `@return tag should be present and have description`. However, the need to have description would not seem to be correct if you would change the regexp the way you propose. Shouldn't the message be changed somehow as well?

**Comment by nmancus1** [comment]
> Shouldn't the message be changed somehow as well?

Good point. We would need to add another message, something like `@return tag should have description`, which would then allow you to filter by message.

We could use `WriteTag` for your purpose, if it logged violations from AST, but it does not:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="SuppressionSingleFilter">
        <property name="checks" value="WriteTag"/>
        <property name="message" value="Javadoc tag @return="/>
    </module>
    <module name="TreeWalker">
        <module name="WriteTag">
            <property name="tag" value="@return"/>
            <property name="tokens"
                      value="METHOD_DEF"/>
            <property name="tagFormat" value="[\s\S]*"/>
            <property name="tagSeverity" value="error"/>
            <message key="type.missingTag"
                     value="Method Javadoc comment is missing @return tag."
            />
        </module>

<!--        This suppression won't work because WriteTag doesn't log errors via AST.-->
        <module name="SuppressionXpathSingleFilter">
            <property name="checks" value="WriteTag"/>
            <property name="query" value="not //METHOD_DEF/TYPE/LITERAL_VOID"/>
        </module>

        <module name="JavadocMethod">
            <property name="allowMissingParamTags" value="false"/>
            <property name="allowMissingReturnTag" value="true"/>
        </module>
    </module>
</module>


```

So this config still reports missing return tag on `void` methods...

Just curious, why require `@return` tag at all if it can be empty?

**Comment by tosix1988** [comment]
It's just an internal requirement of ours to have javadocs with all tags (@param and @return) present, with their descriptions being optional.

**Comment by nmancus1** [comment]
@pbludov maybe this is a chance to begin replacing JavadocMethod with more granular checks... I propose we introduce `JavadocReturnTagCheck` which can check for the presence of `@return` tags for non-void methods, and their respective content? In regards to https://github.com/checkstyle/checkstyle/issues/9745, we could include support for inline tags in this check, as well. As a migration note, we can simply tell `JavadocMethod` users to use `allowMissingReturnTags` property, and add this new check to their config.

**Comment by pbludov** [comment]
@nmancus1 you are right. If we need to add new functionality, it is better to do it in a new check.
Since we got the Javadoc parser, parsing Javadoc with regular expressions doesn't make sense. But the implementation of AbstractJavadocCheck is not yet complete. Issue #5770 should be done to allow xpath suppressions.

**Comment by romani** [comment]
allowance of empty param is bug.
violation on empty return tag is expected , by design. 
If do want to avoid violation it is better to not use Check at all.
but you still what all except for this specific violation you can suppress violation by SuppressionXpathSingleFilter as violation is placed on method not inside javadoc:
```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="allowMissingParamTags" value="false"/>
            <property name="allowMissingReturnTag" value="false"/>
        </module>
        <module name="SuppressionXpathSingleFilter">
          <property name="checks" value="JavadocMethod"/>
          <property name="message" value="@return tag should be present and have description"/>
        </module>
    </module>
</module>

$ cat Test.java 
public class Test {

  /**
   * Returns the sum of the two numbers, {@code a} and {@code b}.
   *
   * @param a first summand
   * @param b
   * @return
   */
  public int sum(int a, int b) {
    return a + b;
  }

  /**
   * Returns the product of two numbers, {@code a} and {@code b}.
   *
   * @param a first multiplicand
   * @param b
   * @return product of {@code a} and {@code b}
   */
  public int prod(int a, int b) {
    return a * b;
  }

}

$ java -jar checkstyle-8.42-all.jar -c config.xml  Test.java
Starting audit...
Audit done.

```
